src/qibosoq/programs/base.py

```diff
@@ -328,28 +343,24 @@ def add_muxed_readout_to_register(self, ro_pulses: List[Pulse]):
 
         self.set_pulse_registers(ch=gen_ch, style="const", length=length, mask=mask)
 
-    def create_mux_ro_dict(self) -> dict:
-        """Create a dictionary containing grouped readout pulses.
+    def group_mux_ro(self) -> list:
+        """Create a list containing readout pulses grouped by start time.
 
-        Example of dictionary:
-        { 'start_time_0': [pulse1, pulse2],
-        'start_time_1': [pulse3]}
+        Example of list:
+        [[pulse1, pulse2], [pulse3]]
         """
-        mux_dict = {}
+        mux_list = []
+        len_last_readout = 0
         for pulse in (pulse for pulse in self.sequence if pulse.type == "readout"):
-            if round(pulse.start, 5) not in mux_dict:
-                if len(mux_dict) > 0:
-                    if (pulse.start - list(mux_dict)[-1]) < 2:  # TODO not 2, but pulses len
-                        mux_dict[round(pulse.start, 5)] = mux_dict[list(mux_dict)[-1]]
-                        del mux_dict[list(mux_dict)[-2]]
-                    else:
-                        mux_dict[round(pulse.start, 5)] = []
-                else:
-                    mux_dict[round(pulse.start, 5)] = []
-            mux_dict[round(pulse.start, 5)].append(pulse)
-
-        self.readouts_per_experiment = len(mux_dict)
-        return mux_dict
+            if pulse.start_delay <= len_last_readout and len(mux_list) > 0:
+                # add the pulse to the last multiplexed readout
+                mux_list[mux_list[-1]].append(pulse)
```

Comment:
This is definitely weird...

Which is the type of `mux_list` elements?

It should be `mux_list: list[int]`, to use `mux_list[-1]` as an index, but it should be `mux_list: list[list[Pulse]]` to call `mux_list[...].append(pulse)`...